torch/nn/modules/transformer.py

```diff
@@ -123,13 +123,20 @@ def forward(self, src, tgt, src_mask=None, tgt_mask=None,
                               memory_key_padding_mask=memory_key_padding_mask)
         return output
 
-    def generate_square_subsequent_mask(self, sz):
+    @staticmethod
+    def generate_square_subsequent_mask(sz, device=None):
         r"""Generate a square mask for the sequence. The masked positions are filled with float('-inf').
             Unmasked positions are filled with float(0.0).
+
+            tensor([[0., -inf, -inf,  ..., -inf, -inf, -inf],
+                    [0.,   0., -inf,  ..., -inf, -inf, -inf],
+                    [0.,   0.,   0.,  ..., -inf, -inf, -inf],
+                    ...,
+                    [0.,   0.,   0.,  ...,   0., -inf, -inf],
+                    [0.,   0.,   0.,  ...,   0.,   0., -inf],
+                    [0.,   0.,   0.,  ...,   0.,   0.,   0.]])
         """
-        mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
-        mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
-        return mask
+        return torch.tensor(float('-inf'), device=device).repeat(sz, sz).triu(1)
```

Comment:
If goal of `torch.tensor(float('-inf'), device=device).repeat(sz, sz)` to create a square `sz` by `sz` matrix filled with `-inf`, you could use `torch.full((sz, sz), float('-inf'), device = device)`. Alternatively you can also try replacing `repeat(sz, sz)` by `expand(sz, sz)` (which will not allocate a square matrix before actually doing `triu`)

Reply:
Using `expand` seems much more productive. Thank you, this is a very useful comment!

Reply:
```torch.triu(torch.full((size, size), float('-inf'), device=device), 1)```
is even faster:

```
mask1 took 0.000079
mask2 took 0.000015
speedup: 5.166978
```

Speed was compared using:
```
import torch
import time

device = 'cuda:0'
size = 1000
n_repeats = 100

# call both mask functions once to prevent warmup delays
torch.tensor(float('-inf'), device=device).repeat(size, size).triu(1)
torch.triu(torch.full((size, size), float('-inf')), 1)

time_sum1 = 0
time_sum2 = 0
for i in range(n_repeats):
    # measure time for mask 1
    start = time.time()
    torch.tensor(float('-inf'), device=device).repeat(size, size).triu(1)
    time_sum1 += time.time() - start
    # measure time for mask 2
    start = time.time()
    torch.triu(torch.full((size, size), float('-inf'), device=device), 1)
    time_sum2 += time.time() - start

mask1_time = time_sum1 / n_repeats
mask2_time = time_sum2 / n_repeats
speedup = mask1_time / mask2_time
print(f'mask1 took {mask1_time:f}')
print(f'mask2 took {mask2_time:f}')
print(f'speedup: {speedup :f}')
```

Reply:
As @vadimkantorov proposed using ```torch.full((sz, sz), float('-inf'), device = device)``` would be a good idea.
This is about 4 times faster than with ```expand(sz, sz)```

Compare using this code:
```
import torch
import time

device = 'cuda:0'
size = 1000
n_repeats = 100

torch.tensor(float('-inf'), device=device).expand(size, size).triu(1)
torch.triu(torch.full((size, size), float('-inf')), 1)

time_sum1 = 0
time_sum2 = 0
for i in range(n_repeats):
    # measure time for mask 1
    start = time.time()
    torch.tensor(float('-inf'), device=device).expand(size, size).triu(1)
    time_sum1 += time.time() - start
    # measure time for mask 2
    start = time.time()
    torch.full((size, size), float('-inf'), device=device).triu(1)
    time_sum2 += time.time() - start

mask1_time = time_sum1 / n_repeats
mask2_time = time_sum2 / n_repeats
speedup = mask1_time / mask2_time
print(f'mask1 took {mask1_time:f}')
print(f'mask2 took {mask2_time:f}')
print(f'speedup: {speedup :f}')
```